backend/pkg/redis/ratelimit.go

```diff
@@ -0,0 +1,125 @@
+package redis
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/redis/go-redis/v9"
+
+	"github.com/mendersoftware/mender-server/pkg/rate"
+)
+
+func NewFixedWindowRateLimiter(
+	client Client,
+	paramsFromContext RatelimitParamsFunc,
+) rate.Limiter {
+	return &fixedWindowRatelimiter{
+		client:     client,
+		paramsFunc: paramsFromContext,
+		nowFunc:    time.Now,
+	}
+}
+
+type RatelimitParams struct {
+	Burst     uint64
+	Interval  time.Duration
+	KeyPrefix string
+}
+
+type RatelimitParamsFunc func(context.Context) (*RatelimitParams, error)
+
+func FixedRatelimitParams(params RatelimitParams) RatelimitParamsFunc {
+	return func(ctx context.Context) (*RatelimitParams, error) {
+		return &params, nil
+	}
+}
+
+type fixedWindowRatelimiter struct {
+	client     Client
+	paramsFunc RatelimitParamsFunc
+	nowFunc    func() time.Time
+}
+
+type simpleReservation struct {
+	ok     bool
+	tokens uint64
+	delay  time.Duration
+}
+
+func (r *simpleReservation) OK() bool {
+	return r.ok
+}
+
+func (r *simpleReservation) Delay() time.Duration {
+	return r.delay
+}
+
+func (r *simpleReservation) Tokens() uint64 {
+	return r.tokens
+}
+
+func epoch(t time.Time, interval time.Duration) int64 {
+	return t.UnixMilli() / interval.Milliseconds()
+}
+
+func fixedWindowKey(prefix string, epoch int64) string {
+	if prefix == "" {
+		prefix = "ratelimit"
+	}
+	return fmt.Sprintf("%s:e:%d:c", prefix, epoch)
+}
+
+func (rl *fixedWindowRatelimiter) Reserve(ctx context.Context) (rate.Reservation, error) {
+	now := rl.nowFunc()
+	params, err := rl.paramsFunc(ctx)
+	if err != nil {
+		return nil, err
+	}
+	epoch := epoch(now, params.Interval)
```

Comment:
In case there is no limit GetLimit will return 0,  RateLimitsFromContext will return -1 as limit, NewFixedWindowRateLimiter will return nil, nil and the code will crash here. Tested on my local setup.

Reply:
I covered the case where the params and error is nil as "no limit" returning a trivial reservation.